src/Agent.Worker/Build/TfsVCTeeSourceProvider.cs

```diff
@@ -0,0 +1,78 @@
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Newtonsoft.Json;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Build
+{
+    public sealed class TfsVCTeeSourceProvider : SourceProvider, ISourceProvider
+    {
+        public override string RepositoryType => WellKnownRepositoryTypes.TfsVersionControl;
+
+        public async Task GetSourceAsync(
+            IExecutionContext executionContext,
+            ServiceEndpoint endpoint,
+            TrackingConfig trackingConfig,
+            CancellationToken cancellationToken)
+        {
+            // Validate args.
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNull(endpoint, nameof(endpoint));
+            ArgUtil.NotNull(trackingConfig, nameof(trackingConfig));
+
+            // Create the tf command manager.
+            var tf = HostContext.CreateService<ITeeTFCommandManager>();
+            tf.CancellationToken = cancellationToken;
+            tf.Endpoint = endpoint;
+            tf.ExecutionContext = executionContext;
+
+            // Create the workspace.
+            AgentSettings settings = HostContext.GetService<IConfigurationStore>().GetSettings();
+            string workDirectory = IOUtil.GetWorkPath(HostContext);
+            string workspace = $"ws_{trackingConfig.BuildDirectory}_{settings.AgentId}";
```

Comment:
Two reasons using the trackingConfig is better than relying on the Variables dictionary:
1) It's not a property bag!
2) This design will work with future plans to sync multiple repos.